add exclude-regex-patterns

jpaskhay opened the pull request:
This mimics the functionality of the `exclude-entropy-patterns` option, but for regex scans. It is heavily based off #192 and other subsequent work around that feature. This will help reduce false positives such as env variables for user info auth in URLs.

Note that the `scope` is forced to `line` and not configurable due to the way the regex scan is done compared to the entropy scan.

A few unrelated formatting changes were necessary for `pre-commit` linters to pass.

I ran into issues trying to get the development environment setup properly (mainly poetry version differences). The new tests that were added passed, but I was running into 2 unrelated test failures -- not sure if related to the dev setup issues.

To help us get this pull request reviewed and merged quickly, please be sure to include the following items:

* [x] Tests (if applicable)
* [x] Documentation (if applicable)
* [x] Changelog entry
* [x] A full explanation here in the PR description of the work done

## PR Type
What kind of change does this PR introduce?

* [ ] Bugfix
* [x] Feature
* [x] Code style update (formatting, local variables)
* [ ] Refactoring (no functional changes, no api changes)
* [ ] Build related changes
* [ ] CI related changes
* [x] Documentation content changes
* [x] Tests
* [ ] Other

## Backward Compatibility

Is this change backward compatible with the most recently released version? Does it introduce changes which might change the user experience in any way? Does it alter the API in any way?

* [x] Yes (backward compatible)
* [ ] No (breaking changes)


## Issue Linking
<!--
    KEYWORD #ISSUE-NUMBER
    [closes|fixes|resolves] #
-->

## What's new?
- add `exclude-regex-patterns` option

jpaskhay committed:
add exclude-regex-patterns

- mimics the functionality of the exclude-entropy-patterns option, but for regex scans. this will help reduce false positives such as env variables for user info auth in URLs
- note that the scope is forced to "line" and not configurable due to the way the regex scan is done compared to the entropy scan
- will follow up with changelog update after creating PR
- a few unrelated changes were necessary pre-commit linters to pass

jpaskhay committed:
update changelog with `exclude-regex-patterns`

jpaskhay commented:
welp, guess that unit test failure wasn't related to my dev environment 😄 . will troubleshoot a bit more and hopefully have a fix soon

rbailey-godaddy commented on docs/source/configuration.rst:
```diff
@@ -203,6 +203,53 @@ match-type   No       String ("search" or "match")  Whether to perform a `search
+.. regex-exclusion-patterns:
+
+Regex Exclusion Patterns
+++++++++++++++++++++++++++
```
```suggestion
Regex Exclusion Patterns
++++++++++++++++++++++++
```
I think the underline length needs to match the header

rbailey-godaddy commented on docs/source/configuration.rst:
```diff
@@ -203,6 +203,53 @@ match-type   No       String ("search" or "match")  Whether to perform a `search
+reason       No       String                       A plaintext reason the exclusion has been added
+match-type   No       String ("search" or "match")  Whether to perform a `search or match`_ regex operation
+============ ======== ============================ ==============================================================
+
```
You addressed this in the PR description, but it would be worth throwing in a line something like "The pattern is always tested against the entire line." (Or any other words that would explain why `scope` is a thing just above for entropy exclusions, but isn't here.)

jpaskhay replied:
sure. will do

rbailey-godaddy commented on tartufo/config.py:
```diff
@@ -278,12 +278,15 @@ def compile_rules(patterns: Iterable[Dict[str, str]]) -> List[Rule]:
+                    f"Invalid value for scope: {pattern.get('scope')}"
+                ) from exc
+        else:
+            scope = Scope.Line
```
This would be more direct this way:
```suggestion
        if exclude_type == "regex":
            # regex exclusions always have line scope
            scope = Scope.Line
        else:
            # entropy exclusions can specify scope
            try:
                scope = Scope(pattern.get("scope", Scope.Line.value))
            except ValueError as exc:
                raise ConfigException(
                    f"Invalid value for scope: {pattern.get('scope')}"
                ) from exc
```

jpaskhay replied:
good catch, agree here. ended up focusing on the unit tests that i didn't self-scrutinize the code changes as much as i should have

rbailey-godaddy commented on tartufo/util.py:
```diff
@@ -242,7 +257,7 @@ def _style_func(msg: str, *_: Any, **__: Any) -> str:
 
 
-def fail(msg: str, ctx: click.Context, code: int = 1) -> NoReturn:
+def fail(msg: str, ctx: click.Context, code: int = 1) -> None:
```
I haven't gotten to looking at the unit test failures yet, but I was surprised to see this return type change. What is the motivation for this? Isn't it still true that `fail()` never returns control to the caller?

jpaskhay replied:
i was getting a `mypy` failure on the pre-commit hook with:

```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1
tartufo/util.py:261: error: Implicit return in function which does not return
```
i can try reverting this to see if it complains in CICD. this change as is has caused another side-effect linting issue, which i was surprised wasn't flagged locally in pre-commit

rbailey-godaddy replied:
I think the mypy failure you're getting now is due to the return-value change:

```
tartufo/commands/update_signatures.py:273: error: Incompatible return value type (got "Optional[GitRepoScanner]", expected "GitRepoScanner")  [return-value]
```
It used to be that `scanner` was known not to be `None` by L273 because of the test at L250 combined with the knowledge that control would not return from the call to `util.fail()` on L252. Now mypy assumes control will return and that consequently `scanner` could still be `None` -- thus the error.

jpaskhay replied:
yup, exactly. was just posting the previous mypy failure which led me to the `NoReturn` -> `None` change. wrapping up wordsmithing the documentation change and i'll push up a revert for this line along with it.

rbailey-godaddy reviewed (changes requested):
This generally looks pretty good - I have some nitpicking below and then there's the question of the failing CI tests, which we'll look at next...

jpaskhay committed:
Apply suggestions from code review

Co-authored-by: Scott Bailey <72747501+rbailey-godaddy@users.noreply.github.com>

jpaskhay committed:
update documentation to note lack of scope option, attempting to revert type hint change

jpaskhay commented:
i'm at a loss on the unit test failure. i did some troubleshooting locally, and the `blake2s` isn't getting mocked, so the real thing is getting called

the way it is getting patched seems proper (same w/ the various other patches in `test_util.py`). not seeing anything obvious that would've changed that in the change set 🤔 . even tried adding a `importlib.reload(util)`

rscottbailey committed:
Fix unit test

Force cache flush before calling function with @lru_cache decorator;
this ensures the return actually reflects the environment constructed by
the unit test and not something left over from an unrelated test.

rscottbailey committed:
Merge branch 'main' into add_exclude_regex_patterns
